File: cc/greedy/NumberOfLinesToWriteString.cc
#include <vector>

using namespace std;

/**
 * LeetCode 806. Number of Lines To Write String
 * https://leetcode.com/problems/number-of-lines-to-write-string/
 */
class NumberOfLinesToWriteString {
public:
    vector<int> numberOfLines(vector<int>& widths, string s) {
        vector<int> r(2, 0); // [lineNum, lastLinePixels]
        int maxPixelsPerLine = 100;
        int n = s.size();
        for (int i = 0; i < n; /* ... */) {
            int lastLinePixels = 0;
            while (i < n && lastLinePixels + widths[s[i] - 'a'] <= maxPixelsPerLine) {
                lastLinePixels += widths[s[i] - 'a'];
                i++;
            }
            r[0]++;
            r[1] = lastLinePixels;
        }
        return r;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}

/* EOF */

File: java/src/io/zihengcat/leetcode/greedy/NumberOfLinesToWriteString.java
package io.zihengcat.leetcode.greedy;

/**
 * LeetCode 806. Number of Lines To Write String
 * https://leetcode.com/problems/number-of-lines-to-write-string/
 */
public class NumberOfLinesToWriteString {
    public static void main(String[] args) {
        // ...
    }
    public int[] numberOfLines(int[] widths, String s) {
        if (widths == null || s == null) {
            return null;
        }
        int lineCnt = 0;
        int lastLinePixels = 0;
        int n = s.length();
        for (int i = 0; i < n; /* ... */ ) {
            int cLinePixels = 0;
            while (i < n && cLinePixels + widths[s.charAt(i) - 'a'] <= 100) {
                cLinePixels += widths[s.charAt(i) - 'a'];
                i++;
            }
            lineCnt++;
            lastLinePixels = cLinePixels;
        }
        return new int[] {
            lineCnt,
            lastLinePixels
        };
    }
}

/* EOF */
